tools/dds/dds-server/lrs-device-watcher.cpp

```diff
@@ -0,0 +1,76 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2022 Intel Corporation. All Rights Reserved.
+
+#include <algorithm>
+#include "lrs-device-watcher.h"
+
+using namespace tools;
+
+lrs_device_watcher::lrs_device_watcher()
+    : _running( false )
+    , _ctx( "{"
+            "\"dds-discovery\" : false"
+            "}" )
+{
+}
+
+
+lrs_device_watcher::~lrs_device_watcher() 
+{
+    _running = false;
+}
+
+void lrs_device_watcher::run( std::function< void( rs2::device ) > add_device_cb,
+                              std::function< void( rs2::device ) > remove_device_cb )
+{
+    notify_connected_devices_on_wake_up( add_device_cb );
+
+    // Register to LRS device change callback.
+    // For each device added, call callback and store it in a list
+    // For each device removed, call callback and remove it from the list
+    _ctx.set_devices_changed_callback(
+        [this, add_device_cb, remove_device_cb]( rs2::event_information & info ) {
```

Comment:
It seems possible that this callback will persist past the lifetime of the watcher...?

Reply:
All of them are destroyed 1 by 1 when the rs-dds-server tool is destroyed. (CTRL + C hit)
I tried adding an extra protection by a  `_running` check,
You think it's enough?

Or should I add a `weak_ptr` of `this` or something like that?

Reply:
I added a weak_ptr for this case, do you think I should do the same for the broadcaster?
```
dev_watcher.run( [&]( rs2::device dev ) { broadcaster.add_device( dev ); },
                     [&]( rs2::device dev ) { broadcaster.remove_device( dev ); } );
```

Reply:
Still not good, let's talk about it

Reply:
Changed as discussed